plugins/nf-azure/src/main/nextflow/cloud/azure/batch/AzBatchProcessObserver.groovy

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2025, Seqera
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package nextflow.cloud.azure.batch
+
+import groovy.transform.CompileStatic
+import groovy.util.logging.Slf4j
+import nextflow.Session
+import nextflow.processor.TaskProcessor
+import nextflow.trace.TraceObserver
+
+/**
+ * Observer that handles process termination events for Azure Batch executor.
+ * When a process terminates (all tasks have been submitted), this observer 
+ * will eagerly set the corresponding Azure Batch job to auto-terminate when
+ * all tasks complete.
+ *
+ * @author Adam Talbot <adam.talbot@seqera.io>
+ */
+@Slf4j
+@CompileStatic
+class AzBatchProcessObserver implements TraceObserver {
+
+    AzBatchProcessObserver(Session session) {
+        // Session not needed, but required by factory interface
+    }
+
+    /**
+     * Called when a process terminates (all tasks have been submitted).
+     * Sets Azure Batch jobs to auto-terminate when all tasks complete.
+     */
+    @Override
+    void onProcessTerminate(TaskProcessor processor) {
```

Comment:
Why using an observer instead of having this logic in the task hander? 

https://github.com/nextflow-io/nextflow/blob/08c0027038cd99789dd790f2082f50870bc2de91/plugins/nf-azure/src/main/nextflow/cloud/azure/batch/AzBatchTaskHandler.groovy#L127-L127

Reply:
Fair, but I still don't see a big value using the trace observer. Would not make more sense to keep this in the cleanup logic here 

https://github.com/nextflow-io/nextflow/blob/5839_azure_batch_jobs_terminate_upon_completion/plugins/nf-azure/src/main/nextflow/cloud/azure/batch/AzBatchService.groovy#L1065-L1080


all related metadata should be accessible in the AzBatchService object